### PR TITLE
Bugfix for remote linked multimedia that already exists downstream

### PR DESCRIPTION
**Pull request by nospame**
## Product Description
Fixes a bug that prevented updates of some linked apps due to multimedia conflicts.

## Technical Summary
[USH-2664](https://dimagi-dev.atlassian.net/browse/USH-2664)
Previously, any `CommCareMultimedia` pulled from a remote, upstream app would always be assigned the id of the upstream `CommCareMultimedia`. The file data is then created or accessed locally based on its hash and attached to the multimedia object. In the latter case, where the same file already exists locally, this would cause unpredictable behavior, with the most common outcome being a reference in the downstream app's `multimedia_map` to a `CommCareMultimedia` object that doesn't exist downstream. In local testing I found that this would also sometimes change the id of the existing `CommCareMultimedia` or create a new multimedia object using the upstream id, but attempting to fetch or store the file data based on its hash would result in a `ResourceConflict` or `ResourceNotFound`.

This change modifies the function that fetches upstream data to:

- use the downstream `CommCareMultimedia` id in the `multimedia_map` if the file already exists (based on its hash)
- save the original upstream id in the multimedia map as `upstream_media_id`
- use the `upstream_media_id`, if it exists, to make the remote request to upstream to fetch file data

In cases where there the `CommCareMultimedia` is newly created (i.e. it has no id) or where the media map item already has an `upstream_media_id` assigned, the previous behavior of copying the upstream id will still apply.

### Safety story
The reassignment of multimedia map ids is limited to the particular case described above: where the same media already exists locally. Tested various situations locally (new media conflict, existing media conflict, pulling via update app/sync media/force pull media) to work through different scenarios.

### Automated test coverage

Added automated test under `TestRemoteLinkedApps` for new functionality.

### QA Plan
[QA-4952](https://dimagi-dev.atlassian.net/browse/QA-4952)
Plan to send to QA in case there are other situations or app config options I haven't considered, or if for some reason this acts differently outside a local environment.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

[USH-2664]: https://dimagi-dev.atlassian.net/browse/USH-2664?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[QA-4952]: https://dimagi-dev.atlassian.net/browse/QA-4952?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: corehq/apps/linked_domain/util.py [hunk 1]
```diff
@@ -102,6 +102,8 @@ def pull_missing_multimedia_for_app(app, old_multimedia_ids=None, force=False):
         media_to_pull = _get_missing_multimedia(app, old_multimedia_ids)
     remote_details = app.domain_link.remote_details
     fetch_remote_media(app.domain, media_to_pull, remote_details)
+    if force:
+        app.save()
```
Comment (nospame) [inline corehq/apps/linked_domain/util.py:106]: I don't love this, but it gives an easy way to fix apps that already have multimedia broken due to local conflict. As a user, I think I would assume that to "force pull" multimedia might also rewrite (and save) any broken multimedia references, which is what this is doing.

File: corehq/apps/linked_domain/tests/test_linked_apps.py [hunk 2]
```diff
@@ -638,6 +638,31 @@ def test_fetch_missing_media(self):
         content = media.fetch_attachment(list(media.blobs.keys())[0])
         self.assertEqual(data, content)
 
+    def test_fetch_missing_media_already_exists(self):
```
Comment (snopoke) [inline corehq/apps/linked_domain/tests/test_linked_apps.py:641]: some comments in this test would be useful to make it easier to understand what different steps are for

File: corehq/apps/app_manager/views/utils.py [hunk 3]
```diff
@@ -352,7 +352,10 @@ def update_linked_app(app, master_app_id_or_build, user_id):
         or master_build.version > previous.upstream_version
         or toggles.MULTI_MASTER_LINKED_DOMAINS.enabled(app.domain)
     ):
-        old_multimedia_ids = set([media_info.multimedia_id for path, media_info in app.multimedia_map.items()])
+        old_multimedia_ids = set(
+            [getattr(media_info, 'upstream_media_id', None) or media_info.multimedia_id
+             for path, media_info in app.multimedia_map.items()]
+        )
```
Comment (snopoke) [inline corehq/apps/app_manager/views/utils.py:358]: Why is this change necessary? I think `old_multimedia_ids` is only used to determine which media is missing from the local server in which case it should only need the local ID and not the upstream ID. I think using the upstream ID will create false positives for missing media.

**Review by snopoke — COMMENTED** [review]
Generally looks good. One main comment.

File: corehq/apps/app_manager/views/utils.py (hunk 3 above)
Comment (nospame) [reply corehq/apps/app_manager/views/utils.py:358]: I found that "old" here means the collection of multimedia already present in the older version of the downstream app, before the newer app version is pulled. [_get_missing_multimedia](https://github.com/dimagi/commcare-hq/pull/32797/files#diff-b565707cd8670d0f76525035ea595b0d2617a8d87e5575e5273d945f33e6468eR123-R135) ultimately checks the ids from the newly pulled multimedia map against `old_multimedia_ids`, and if anything is missing from `old_multimedia_ids`, adds it to a list to `fetch_remote_media`. 

Essentially I made this change _because_ I was getting false positives when looking for updated multimedia ids in a list containing only the local ids, resulting in a new `AuxMedia` entry on every app pull.

Comment (nospame) [reply corehq/apps/app_manager/views/utils.py:358]: Well, you were correct that this did have an unintended consequence. On repeated app pulls (particularly with the `MULTI_MASTER` feature flag enabled) the multimedia map is not being rewritten from its upstream version, because that happens in `fetch_remote_media` which is never called if the multimedia does not appear to be "missing" in the first place.

Reverting this line is a clear way to fix the issue, but then a new `AuxMedia` entry gets created on the `CommCareMultimedia` object on every app pull (where there is media that conflicts between upstream & downstream). I assume this isn't ideal, but how big a concern would it be, really?

File: corehq/apps/linked_domain/remote_accessors.py [hunk 4]
```diff
@@ -126,7 +126,8 @@ def fetch_remote_media(local_domain, missing_media, remote_app_details):
             item['multimedia_id'], item['upstream_media_id'] = media_item._id, item['multimedia_id']
         else:
             media_item._id = item['multimedia_id']
-        media_item.attach_data(content, original_filename=filename)
+        if not any(a in media_item.aux_media for a in media_item.aux_media if a.uploaded_filename == filename):
```
Comment (nospame) [inline corehq/apps/linked_domain/remote_accessors.py:129]: I *think* this is the check I need to make here, since it seems like `AuxMedia` exists just to ensure the app's reference to the filename is valid.

Also thought about extracting this line along with 124 into named functions for clarity (probably `_media_exists_locally` and `_needs_data_attached`, but I wasn't sure if it would be helpful or needed for the one-liners that they are.

File: corehq/apps/app_manager/views/utils.py (hunk 3 above)
Comment (snopoke) [reply corehq/apps/app_manager/views/utils.py]: I looked at this again and think your original change was correct in terms of pulling missing MM. The issue you describe with not updating the multimedia map seems like a separate issue.

I think we should look at updating the `corehq.apps.app_manager.views.utils.overwrite_app` method to do a smarter conversion of the MM map rather than just overwriting the local version with the remote version. Bascially we want to preserve the 'upstream_id' in the media items. I think then we'd be able to correctly determine what's actually missing without having to re-fetch anything.

Comment (nospame) [reply corehq/apps/app_manager/views/utils.py:365]: @snopoke I took a stab at this today, let me know if it looks right. Now, it's still `fetch_remote_media` that's responsible to set the correct ids for the local and the upstream reference, because it uses the actual file data/hash to determine the media already exists locally. `overwrite_app` is now also comparing the existing and incoming multimedia maps, and if an item in the existing map has the same file path plus an `upstream_media_id` that matches the incoming `multimedia_id`, it keeps the existing multimedia map item instead. 

There's one minor refactor included as well, setting the incoming multimedia map's versions to `None` during comparison rather than in a separate loop.

File: corehq/apps/linked_domain/remote_accessors.py [hunk 5]
```diff
@@ -121,13 +121,18 @@ def fetch_remote_media(local_domain, missing_media, remote_app_details):
         media_class = CommCareMultimedia.get_doc_class(item['media_type'])
         content = _fetch_remote_media_content(item, remote_app_details)
         media_item = media_class.get_by_data(content)
-        media_item._id = item['multimedia_id']
+        if getattr(media_item, '_id', None) and not getattr(item, 'upstream_media_id', None):
```
Comment (snopoke) [inline corehq/apps/linked_domain/remote_accessors.py:124]: `and not getattr(item, 'upstream_media_id', None)`

Just noting that this condition shouldn't be necessary since if the `item` did have an `upstream_media_id` it indicates that the media exists locally already and it shouldn't be in the list of `missing_media`.

Fine to keep this condition as a defensive guard but we should also be testing that the `_get_missing_multimedia` is working correctly and not returning media that exists locally.

File: corehq/apps/app_manager/views/utils.py [hunk 6]
```diff
@@ -197,14 +198,20 @@ def overwrite_app(app, master_build, report_map=None):
...
-    for path in wrapped_app.multimedia_map.keys():
-        wrapped_app.multimedia_map[path].version = None
     wrapped_app.set_media_versions()
 
     enable_usercase_if_necessary(wrapped_app)
     return wrapped_app
 
 
+def _update_multimedia_map(old_map, new_map):
+    for path in new_map:
```
Comment (snopoke) [inline corehq/apps/app_manager/views/utils.py:208]: I'm not sure if this scenario is possible but we should make sure that none of the items in 'new_map' have 'upstream_media_id' attributes (or at least we should remove them).

File: corehq/apps/app_manager/views/utils.py [hunk 7]
```diff
@@ -352,7 +359,10 @@ def update_linked_app(app, master_app_id_or_build, user_id):
         or master_build.version > previous.upstream_version
         or toggles.MULTI_MASTER_LINKED_DOMAINS.enabled(app.domain)
     ):
-        old_multimedia_ids = set([media_info.multimedia_id for path, media_info in app.multimedia_map.items()])
+        old_multimedia_ids = set(
+            [getattr(media_info, 'upstream_media_id', None) or media_info.multimedia_id
```
[the remark this answers is not in the document]
Comment (nospame) [reply corehq/apps/app_manager/views/utils.py]: I found there didn't seem to be any testing added for what `old_multimedia_ids` should really be, so I added a few test cases.

File: corehq/apps/linked_domain/remote_accessors.py [hunk 8]
```diff
@@ -121,13 +121,18 @@ def fetch_remote_media(local_domain, missing_media, remote_app_details):
...
+            # media already exists locally; update the multimedia_map
+            item['multimedia_id'], item['upstream_media_id'] = media_item._id, item['multimedia_id']
+        else:
+            media_item._id = item['multimedia_id']
         media_item.attach_data(content, original_filename=filename)
```
Comment (snopoke) [inline corehq/apps/linked_domain/remote_accessors.py:129]: For media that already exists locally, won't this create a duplicate `AuxMedia`?

**Review by snopoke — APPROVED** [review]
One outstanding question but not blocking. Otherwise I think this is GTG.

File: corehq/apps/linked_domain/remote_accessors.py (hunk 8 above)
Comment (nospame) [reply corehq/apps/linked_domain/remote_accessors.py:129]: Yes, but only once (or on repeated force pulls of multimedia). So for example, a CommCareImage on my local testing app now has two `AuxMedia` entries that are identical except for:
```
"uploaded_filename": "aladdin-sane.jpeg"
"uploaded_filename": "module0_case_search_again_label_item_en.jpeg"
```

Because only the right media is now showing up in `missing_multimedia`, it doesn't get fetched/attached again after the first time. This seems comparable to the behavior I see when uploading the same file to multiple apps locally - a new `AuxMedia` attachment is created once, but in the future the app knows it already exists and doesn't try to add it again.
